src/jesse_validator_draft3.erl

```diff
@@ -765,26 +765,41 @@ check_unique_items(_, false, State) ->
   State;
 check_unique_items([], true, State) ->
     State;
+check_unique_items([_], true, State) ->
+    State;
 check_unique_items(Value, true, State) ->
   try
-    lists:foldl( fun(_Item, []) ->
-                     ok;
-                    (Item, RestItems) ->
-                     lists:foreach( fun(ItemFromRest) ->
-                                        case is_equal(Item, ItemFromRest) of
-                                          true  ->
-                                            throw({?not_unique, Item});
-                                          false -> ok
+%% First we do an efficient check for duplicates: convert the list to a set
+%% and if there are no duplicates, the set and the list have the same length
+%% In order to avoid differences for lists in which order is not relevant
+%% (e.g. JSON properties of an object maybe represented as a proplist), these
+%% lists for which order is not relevant are sorted (objects are normalized).
+%% If the first efficient check fails, then we search for the items that are
+%% duplicated with a less efficient check (that will very seldom be executed).
+    NormalizedValue = jesse_lib:normalize_and_sort(Value),
+    NoDuplicates = sets:from_list(NormalizedValue, [{version, 2}]),
+    case sets:size(NoDuplicates) == length(Value) of
+      true -> State;
+      false ->
+        lists:foldl( fun(_Item, []) ->
+                         ok;
+                        (Item, RestItems) ->
+                         lists:foreach( fun(ItemFromRest) ->
+                                            case is_equal(Item, ItemFromRest) of
+                                              true  ->
+                                                throw({?not_unique, Item});
+                                              false -> ok
+                                            end
                                         end
-                                    end
-                                  , RestItems
-                                  ),
-                     tl(RestItems)
-                 end
-               , tl(Value)
-               , Value
-               ),
-    State
+                                      , RestItems
+                                      ),
+                         tl(RestItems)
+                     end
+                   , tl(Value)
+                   , Value
+                   ),
```

Comment:
since `NormalizedValue` can be directly sorted, this algorithm may be optimized by sorting `lists:sort(NormalizedValue)` (logarithmic) and then you iterate through the sorted list and only compare each element with the next one (linear):

```
find_duplicate([A, A | _]) ->
  throw({?not_unique, A});
find_duplicate([_ | Tail]) ->
  find_duplicate(Tail);
find_duplicate([]) ->
  ok.
```